iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,306 @@
+"""
+Iterative refinement in Bayesian expection maximization setting
+for reconstruction of particles.
+"""
+
+import numpy as np
+from compSPI.transforms import do_fft, do_ifft
+from simSPI.transfer import eval_ctf
+
+
+class IterativeRefinement:
+    """
+    Performs interative refimenent in a Bayesian expectation
+    maximization setting, i.e. maximum a posteriori estimation.
+
+    Parameters
+    ----------
+    map_3d_init : arr
+        Initial particle map.
+        Shape (n_pix, n_pix, n_pix)
+    particles : arr
+        Particles to be reconstructed.
+        Shape (n_particles, n_pix, n_pix)
+    ctf_info : list of dicts
+        Each dict contains CTF k,v pairs per particle.
+            Shape (n_particles,)
+
+    Returns
+    -------
+    map_3d_update : arr
+        Current iteration of map.
+        Shape (n_pix, n_pix, n_pix)
+    map_3d_final : arr
+        Final updated map.
+        Shape (n_pix, n_pix, n_pix)
+    half_map_3d_final_1 : arr
+        Shape (n_pix, n_pix, n_pix)
+    half_map_3d_final_2 : arr
+        Shape (n_pix, n_pix, n_pix)
+    fsc_1d : arr
+        Final one dimensional fourier shell correlation.
+        Shape (n_pix // 2,)
+    """
+
+    def __init__(self, map_3d_init, particles, ctf_info, max_itr=7):
+        self.map_3d_init = map_3d_init
+        self.particles = particles
+        self.ctf_info = ctf_info
+        self.max_itr = max_itr
+
+    def split_array(self, arr):
+        """
+        Split array into two halves along 0th axis.
+
+        Parameters
+        ----------
+        arr : arr
+            Shape (n_pix, n_pix, n_pix)
+
+        Returns
+        -------
+        arr1 : arr
+            Shape (n_pix // 2, n_pix, n_pix)
+        arr2: arr
+            Shape (n_pix // 2, n_pix, n_pix)
+        """
+        idx_half = arr.shape[0] // 2
+        arr_1, arr_2 = arr[:idx_half], arr[idx_half:]
+
+        if arr_1.shape[0] != arr_2.shape[0]:
+            arr_2 = arr[idx_half : 2 * idx_half]
+
+        return arr_1, arr_2
+
+    def build_ctf_array(self):
+        """
+        Build 2D array of evaluated CTFs from inputted
+        CTF parameters for each particle.
+
+        Returns
+        -------
+        ctfs : arr
+            Shape (n_ctfs, n_pix, n_pix)
+        """
+        n_ctfs = len(self.ctf_info)
+        ctfs = []
+
+        for i in range(n_ctfs):
+            ctfs.append(eval_ctf(**self.ctf_info[i]))
```

Comment:
Can you use this CTF instead? 

It's already part of the linear simulator, so it's better to keep the CTF defined in a consistent way. 
https://github.com/compSPI/simSPI/blob/master/simSPI/linear_simulator/ctf.py

You'll just have to do some torch to numpy conversion

Reply:
we're on it @calhep

Reply:
we'll implement this in a future PR and leave this as is for now, and only address docstring/design in this PR.